runtime/near-evm-runner/src/utils.rs

```diff
@@ -183,66 +187,497 @@ pub fn format_log(topics: Vec<H256>, data: &[u8]) -> std::result::Result<Vec<u8>
     Ok(result)
 }
 
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum Type {
+    Address,
+    Uint,
+    Int,
+    String,
+    Bool,
+    Bytes,
+    Byte(u8),
+    Custom(String),
+    Array { length: Option<u64>, inner: Box<Type> },
+}
+
+/// the type string is being validated before it's parsed.
+pub fn parse_type(field_type: &str) -> Result<Type> {
+    #[derive(PartialEq)]
+    enum State {
+        Open,
+        Close,
+    }
+
+    let mut lexer = Lexer::new(field_type);
+    let mut token = None;
+    let mut state = State::Close;
+    let mut array_depth = 0;
+    let mut current_array_length: Option<u64> = None;
+
+    while lexer.token != Token::EndOfProgram {
+        let type_ = match lexer.token {
+            Token::Identifier => Type::Custom(lexer.slice().to_owned()),
+            Token::TypeByte => Type::Byte(lexer.extras.0),
+            Token::TypeBytes => Type::Bytes,
+            Token::TypeBool => Type::Bool,
+            Token::TypeUint => Type::Uint,
+            Token::TypeInt => Type::Int,
+            Token::TypeString => Type::String,
+            Token::TypeAddress => Type::Address,
+            Token::LiteralInteger => {
+                let length = lexer.slice();
+                current_array_length = Some(length.parse().map_err(|_| {
+                    VMLogicError::EvmError(EvmError::InvalidMetaTransactionMethodName)
+                })?);
+                lexer.advance();
+                continue;
+            }
+            Token::BracketOpen if token.is_some() && state == State::Close => {
+                state = State::Open;
+                lexer.advance();
+                continue;
+            }
+            Token::BracketClose if array_depth < 10 => {
+                if state == State::Open && token.is_some() {
+                    let length = current_array_length.take();
+                    state = State::Close;
+                    token = Some(Type::Array {
+                        inner: Box::new(token.expect("if statement checks for some; qed")),
+                        length,
+                    });
+                    lexer.advance();
+                    array_depth += 1;
+                    continue;
+                } else {
+                    return Err(VMLogicError::EvmError(EvmError::InvalidMetaTransactionMethodName));
+                }
+            }
+            Token::BracketClose if array_depth == 10 => {
+                return Err(VMLogicError::EvmError(EvmError::InvalidMetaTransactionMethodName));
+            }
+            _ => {
+                return Err(VMLogicError::EvmError(EvmError::InvalidMetaTransactionMethodName));
+            }
+        };
+
+        token = Some(type_);
+        lexer.advance();
+    }
+
+    Ok(token.ok_or(VMLogicError::EvmError(EvmError::InvalidMetaTransactionMethodName))?)
+}
+
 pub fn near_erc721_domain(chain_id: U256) -> RawU256 {
```

Comment:
Add a comment

Reply:
seem many comments here, i'll address them in a PR